### PR TITLE
Automatic header decryption

### PR DESCRIPTION
**Pull request by hanneary**
# Why
Headers are not being decrypted on requests

# How
Send encrypted headers to e3 along with ciphertexts parsed from the body

### CODE REVIEW

File: data-plane/src/server/layers/decrypt.rs [hunk 1]
```diff
@@ -45,19 +52,26 @@ impl std::convert::From<DecryptError> for Response<Body> {
...
+lazy_static! {
+    static ref CIPHERTEXT_VERSION_REGEX: regex::Regex = get_ciphertext_regex();
+}
+
 #[derive(Clone)]
-pub struct DecryptLayer {
-    e3_client: Arc<E3Client>,
+pub struct DecryptLayer<T: E3Api + Send + Sync + 'static>
+where
+    T: E3Api + Send + Sync + 'static,
```
Comment (doylemark) [inline data-plane/src/server/layers/decrypt.rs:62]: why do we need these bounds beyond E3API?

**Review by doylemark — APPROVED** [review]
very nice! 1 q

File: data-plane/src/server/layers/decrypt.rs [hunk 2]
```diff
@@ -45,19 +52,26 @@ impl std::convert::From<DecryptError> for Response<Body> {
...
 
+lazy_static! {
+    static ref CIPHERTEXT_VERSION_REGEX: regex::Regex = get_ciphertext_regex();
+}
+
 #[derive(Clone)]
-pub struct DecryptLayer {
-    e3_client: Arc<E3Client>,
+pub struct DecryptLayer<T: E3Api + Send + Sync + 'static>
+where
```
Comment (donaltuohy) [inline data-plane/src/server/layers/decrypt.rs:61]: I don't think the `where` with trait bounds is also needed here?

File: data-plane/src/server/layers/decrypt.rs (hunk 1 above)
Comment (hanneary) [reply data-plane/src/server/layers/decrypt.rs]: copy pasta from AuthLayer! have removed from both

**Review by donaltuohy — APPROVED** [review]

File: data-plane/src/server/layers/decrypt.rs (hunk 2 above)
Comment (hanneary) [reply data-plane/src/server/layers/decrypt.rs]: ahh yes removed!

**Pull request merged**